[yeon] 미션4 : 모든 기물 배치하기

kimnayeon0108 opened the pull request:
안녕하세요! 백엔드반 yeon입니다. 미션4 모든 기물 배치하기 구현 완료하고 코드 리뷰 요청드립니다~ 아래에 각 클래스에 대한 설명을 기술하였으니 참고 부탁드립니다

1. Board 클래스
- 리스트 whitePawns와 blackPawns : 기물 pawn을 담는 리스트입니다.
- 리스트 whitePieces와 blackPieces : pawn이외의 기물을 담는 리스트입니다.
- distinguish() : 인자로 받은 piece가 어느 리스트에 들어가야하는지를 판별해주는 기능을 합니다.
- add() : piece를 해당하는 리스트에 추가해주는 기능을 합니다.
- size()와 findPiece() : 이 메소드들은 현재 미션에서는 사용하지 않지만 이전 미션 요구사항에 있었고 다음미션에 사용할 수도 있을거 같아서 메소드명과 매개변수 타입을 수정하고 삭제하지 않고 뒀습니다.
- initialize() : board에 기물들을 초기화하는 기능을 합니다. initBlackPiece(), initBlackPawn(), initWhitePawn(), intitWhitePiece()를 호출합니다.
- initBlackPiece() : pawn을 제외한 black 컬러의 Piece의 객체를 생성하여 리스트에 담습니다.
- initWhitePiece() : pawn을 제외한 white 컬러의 Piece 객체를 생성하여 리스트에 담습니다.
- initWhitePawn() : white의 pawn을 생성하여 리스트에 담습니다.
- initBlackPawn() : black의 pawn을 생성하여 리스트에 담습니다.
- pieceCount() : whitePawns, whitePiece, blackpawns, whitePieces 리스트의 크기의 총합을 구해, 기물들의 총 개수를 반환합니다.
- showBoard() : 리스트에 담긴 기물들의 representation을 포함한 board를 문자열로 반환합니다.
- getEmptyResult() : board의 빈공간을 갖는 한 열을 나타내는 문자열을 반환합니다.
- getWhitePawnResult(), getBlackPawnResult(), getWhitePieceResult(), getBlackPieceResult() : 각각 해당하는 리스트에 담긴 기물들의 representation을 하나의 문자열로 반환합니다.

2. BoardTest 클래스 

- create() : 보드를 초기화 한 후 (board.initialize()) board의 pieceCount()의 리턴값이 예상하는 값(32)와 같은지 확인합니다.

    board.showBoard()의 리턴값이 예상하는 값과 같은지 확인합니다.

3. enum Color와 PieceName

- 체스 기물의 색깔과 종류는 정해져 있기때문에 enum으로 생성했습니다.
- PieceName에 tempRepresentation 이라는 속성으로 각 기물에 해당하는임시의 representation 값을 갖도록 하였습니다.

4. Piece 클래스

- 기존의 Pawn 클래스의 이름을 변경하였습니다.
- Color타입의 color, PieceName 타입의 name, representation 이라는 속서을 갖도록 했습니다.
- 생성자에서 인자로 받은 color와 name을 대입하고, createRepresentation()의 리턴값을 representation에 대입하였습니다. 요구사항에서 private으로 지정하라고 하여 생성자를 private으로 설정하였습니다.
- createRepresentation() : 기물의 name에 해당하는 각 tempRepresentation 값을 기물의 색이 white면 소문자, black이면 대문자로 반환하도록 구현했습니다.
- 각각의 기물을 생성하는 메소드들 createWhitePawn(), createBlackPawn(), createWhiteKnight(), createBlackKnight(), createWhiteRook(), createBlackRook(), createWhiteBishop(), createBlackBishop(), createWhiteQueen(), createBlackQueen(), createWhiteKing(), createBlackKing() : PieceTest의 요구사항에서 기물을 생성하는 메소드를 static 메소드를 호출하도록 되어있어서 static 메소드로 구현했습니다.

    → static메소드여야 하는 이유를 생각해보니, 기물들은 외부에서 생성자를 호출하지 못하도록 private으로 생성자를 설정했고, 이 메소드들이 static이여야 Piece객체 생성없이 호출할수 있기 때문에 static 메소드로 정의되어야 한다고 생각했습니다. 

- isBlack(), isWhite() : 각 기물의 color가 black인지 white인지 여부를 boolean값으로 반환합니다.

5. PieceTest 클래스

- 기존의 PawnTest 의 이름을 변경했습니다.
- create_piece() : 생성된 기물의 생상과 representation이 예상하는 값과 같은지 확인합니다.
- check_color() : Piece클래스의 isBlack()과 isWhte()의 값이 예상하는 값과 같은지 확인합니다.

6. StringUtils 클래스

- StringUtils의 객체를 생성할 수 없도록 기본 생성자를 private으로 설정했습니다.
- new line문자가 운영체제마다 다르다는 것을 보완하기 위해 상수 NEWLINE을 추가하여 System.getProperty("line.separator") 값을 주었습니다.
- appendNewLine : 인자로 받은 문자열에 NEWLINE을 더하여 줄바꿈한 문자열을 반환하도록 하였습니다.


바쁜시간에 코드리뷰 해주셔서 감사합니다🙇‍♀️👏

kimnayeon0108 committed:
Add dependency in  build.gradle

kimnayeon0108 committed:
Add dependency in  build.gradle

kimnayeon0108 committed:
Merge branch 'kimnayeon0108' of https://github.com/codesquad-members-2021/java-chess into kimnayeon0108

kimnayeon0108 committed:
Merge branch 'kimnayeon0108' of https://github.com/kimnayeon0108/java-chess into kimnayeon0108

kimnayeon0108 committed:
Change Class name Pawn to Piece

kimnayeon0108 committed:
Merge branch 'kimnayeon0108' of https://github.com/codesquad-members-2021/java-chess into kimnayeon0108

kimnayeon0108 committed:
Merge branch 'kimnayeon0108' into step4-initialize-all-piece

kimnayeon0108 committed:
Fix Piece class
Remove constructor with no parameter
Add name parameter in constructor

kimnayeon0108 committed:
Fix Piece class
Add enum PieceName
Add createWhitePawn() and createBlackPawn() method

kimnayeon0108 committed:
Add methods that create new instance of each piece using private constructor

kimnayeon0108 committed:
Remove representation field in Color
Add representation field in PieceName

kimnayeon0108 committed:
Add createRepresentation() method
representation has to be different depending on color and name
it has to be upper case when piece is black

kimnayeon0108 committed:
Change variable name representation to tempRepresentation in enum PieceName

kimnayeon0108 committed:
Add getRepresentation() method

kimnayeon0108 committed:
Add create_piece() method that make sure each piece has been created well
fix parameter color type

kimnayeon0108 committed:
Add rep local variable and fix arguments of verifyPiece() method in create_piece() method

kimnayeon0108 committed:
Fix BoardTest class
remove variable white, black cosidered to be useless
fix create() test method according to requirements on lucas
plan to fix Board class according to test method

kimnayeon0108 committed:
Add utils package and add StringUtils.java into chess.utils
make constructor private so that nothing can create StringUtils's instance
add constant NEWLINE for line break
add appendNewLine() method that append parameter line and NEWLINE to global variable 'string'

kimnayeon0108 committed:
Make PieceName class as a file to make it public
seperate from Piece.java file

kimnayeon0108 committed:
Add getName() method so that will be able to get name of Piece instance

kimnayeon0108 committed:
Fix add(), size(), distinguish() method to addPawn(), pawnSize(), distinguishPawn()
since new ArrayList whitePieces and blackPieces are added, methods regarding pawns lists have to be specified that those methods are for List of pawns
add addPiece(), pieceSize(), distinguishPiece() method for pieces with the exception of Pawn, work with whitePieces and blackPieces Lists

kimnayeon0108 committed:
Fix initialize() method calls initBlackPiece(), initBlackPawn(), initWhitePiece(), initWhitePawn()
add initBlackPiece(), initBlackPawn(), initWhitePiece(), initWhitePawn() method

kimnayeon0108 committed:
Combine distinguishPawn() and distinguishPiece() method and rename it as distinguish()

kimnayeon0108 committed:
Remove useless field

kimnayeon0108 committed:
Complete showBoard() method using appendNewLine() util method
add getWhitePieceResult() and getBlackPieceResult()

kimnayeon0108 committed:
Fix code

kimnayeon0108 committed:
Add isWhite(), isBlack() method in Piece class
add check_color() method in PieceTest to  make sure isWhite() and isBlack() method work

kimnayeon0108 committed:
Fix code

kimnayeon0108 committed:
Fix access modifier

kimnayeon0108 committed:
Remove constant COLUMN considered to be useless

honux77 commented:
연 PR에 자세한 설명 달아줘서 고마워요. 그런데 메소드 전체가 일일히 뭐하는지 설명하는 것보다, 내가 무엇을 왜 이렇게 했는지 PR자체의 핵심 내용을 설명해 주면 더 좋을 것 같아요. 너무 빡빡하고 자세해서 지나치게 풍부한 주석같은 느낌을 주네요.

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
+        if (piece.getColor() == Color.BLACK && piece.getName() != PieceName.PAWN) {
+            return blackPieces;
+        }
+        return null;
```
private 메소드는 public 메소드 뒤에 두는 게 좋아요. 관련 내용 검색해 보세요.

honux77 replied:
메소드 이름이나 기능은 좋아요.

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
-        List list = distinguish(pawn);
-        list.add(pawn);
+    public void add(Piece piece) {
+        distinguish(piece).add(piece);
```
아 그런데 요것만 이렇게 보니까? 살짝 뭐하는 거지? 이런 느낌이 들어요. getPieceList(piece) 이런 네이밍이 적절하지 않았을까 싶은 생각이 드네요.

kimnayeon0108 replied:
@honux77 안그래도 조원들끼리 코드리뷰하는 시간에 그얘기가 나왔습니다. 메소드 이름을 바꿔보겠습니다! 
메소드 위치도 public 뒤로 변경하겠습니다 private 메소드가 public 뒤에 나와야하는 이유는 public 메소드 안에서 호출되기 때문에 코드를 읽을때 가독성이 더 좋을거같다라는 생각이 드는데 관련내용은 더 찾아봐야겠습니다

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
 
-    public Pawn findPawn(Pawn pawn, int idx) {
-        return distinguish(pawn).get(idx);
+    public Piece findPiece(Piece piece, int idx) {
```
이거 이상해요. 피스와 인덱스로 다른 피스를 찾아준다?
findPiece(color, name, idx) 가 더 좋아 보여요. 
지금 보니 다른 것들도 그렇게 하는게 더 좋을 것 같기도 하고요.

kimnayeon0108 replied:
@honux77 앗 듣고보니 그렇네요. 수정해보겠습니다!

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
     }
 
     public void initialize() {
+        initBlackPiece();
+        initBlackPawn();
+        initWhitePawn();
+        initWhitePiece();
```
깔끔!

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
+    }
+
+    private void initBlackPiece() {
+        add(Piece.createBlackRook());
+        add(Piece.createBlackKnight());
+        add(Piece.createBlackBishop());
+        add(Piece.createBlackQueen());
+        add(Piece.createBlackKing());
+        add(Piece.createBlackBishop());
+        add(Piece.createBlackKnight());
+        add(Piece.createBlackRook());
```
이것도 괜찮긴 한데 ㅋㅋ
문득
royal()
bishop()
knight()
rook()
요렇게만 호출하면 어떨까? 하는 잡생각이 들었습니다. 신경쓰지 마세요 ㅋㅋ

kimnayeon0108 replied:
@honux77 헉 저도 그생각이 살짝 들긴 했어요...ㅎㅎㅎ

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
+    }
+
+    private void initWhitePawn() {
         for (int i = 0; i < PAWNSNUM; i++) {
```
PAWN_SUM

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
+    }
+
+    public String showBoard() {
+        String empty = appendNewLine(getEmptyResult());
+        return appendNewLine(getBlackPieceResult()) +
+                appendNewLine(getBlackPawnResult()) +
+                empty + empty + empty + empty +
+                appendNewLine(getWhitePawnResult()) +
+                appendNewLine(getWhitePieceResult());
```
보통 스트링 빌더를 많이 쓰긴 했던데 스트링도 나쁘지 않아 보여요. 
혹시 가능하다면 빌더로 바꾸고 재성성된 바이트코드가 어떻게 되는지 확인해 볼 수 있을까요?

kimnayeon0108 replied:
@honux77 다른건 다 StringBuilder로 해놓고 이건 왜 String으로 했는지 모르겠네요...ㅎㅎ 수정해보고 바이트코드 비교 해보겠습니다

kimnayeon0108 replied:
@honux77 StringBuilder로 변경하고 바이트코드를 비교해보니, 위와같이 String +연산자 이용해서 작성한 것의 바이트 코드에도 StringBuilder.append라고  쓰여있습니다. 바이트코드를 처음 보지만 내부적으로 StringBuilder를 사용하는건가? 라는 생각이 들어서 찾아보니 Java 1.5 부터는 내부적으로 StringBuilder로 변환해서 처리한다고 합니다.

kimnayeon0108 replied:
<img width="763" alt="스크린샷 2021-02-19 오후 4 59 13" src="https://user-images.githubusercontent.com/65011131/108496212-1af4d280-72ed-11eb-81dd-25dbecec4df0.png">

honux77 commented on src/main/java/chess/Board.java:
```diff
@@ -1,85 +1,132 @@
+
+    private String getEmptyResult() {
         StringBuilder result = new StringBuilder();
         for (int i = 0; i < PAWNSNUM; i++) {
```
코드를 보니 변수 이름 PAWN_SUM이 약간 어울리지 않아요. BOARD_ROW 이런 걸로 바꾸는게 어때요?

kimnayeon0108 replied:
@honux77 맞습니다. 수정하겠습니다!

honux77 commented on src/main/java/chess/Chess.java:
```diff
@@ -11,7 +11,7 @@ public static void main(String[] args) {
             if (input.equals("y")) {
                 Board board = new Board();
                 board.initialize();
-                System.out.println(board.print());
+                System.out.println(board.showBoard());
             }
             if (input.equals("n")) {
                 System.out.println("프로그램을 종료합니다.");
```
Scanner close가 없군요. try - with 사용해 보세요. 
그리고 yn 외에 다른 문자열 입력에 대한 처리도 없는 것 같아요.

kimnayeon0108 replied:
@honux77 네에 수정해보겠습니다!

honux77 commented on src/main/java/chess/pieces/PieceName.java:
```diff
@@ -0,0 +1,20 @@
+    QUEEN('q'),
+    KING('k');
+
+    private final char tempRepresentation;
```
오오 final! 아니어도 괜찮긴 한데 그래도 명시적인 final도 좋네요.

kimnayeon0108 replied:
@honux77 바뀌지 않아야 하는 값이라서 final 키워드를 붙였는데 어차피 enum 생성자가 private이라서 없어도 괜찮을거 같습니다!

honux77 commented on src/test/java/chess/BoardTest.java:
```diff
@@ -1,46 +1,30 @@
-        board.add(pawn);
-        assertEquals(size, board.size(pawn));
-        assertEquals(pawn, board.findPawn(pawn, idx));
+        assertEquals(32, board.pieceCount());
```
test가 뭘하는 건지 메소드이름을 아주아주 자세히 적어주거나 그렇지 않으면 DisplayName 어노테이션을 활용해 보세요.

kimnayeon0108 replied:
@honux77 네엡 더 생각해보고 수정하겠습니다.

honux77 replied:
앗 그리고 지금 발견했는데 assertEquals() 말고 assertThat() 사용해 주세요! 다음 단계부터!

kimnayeon0108 replied:
@honux77 넵 수정하겠습니다!

honux77 reviewed (changes requested):
사소한 것 말고는 깔끔하게 잘 짰네요! 코멘트 읽어보고 개선한 후에 리뷰 요청해 주세요.

kimnayeon0108 commented:
> 연 PR에 자세한 설명 달아줘서 고마워요. 그런데 메소드 전체가 일일히 뭐하는지 설명하는 것보다, 내가 무엇을 왜 이렇게 했는지 PR자체의 핵심 내용을 설명해 주면 더 좋을 것 같아요. 너무 빡빡하고 자세해서 지나치게 풍부한 주석같은 느낌을 주네요.

네! 피드백 감사합니다.
다음pr부터는 더 간략하게 기술하겠습니다. 
커밋할때도 들었던 생각인데 요구사항대로 구현하다보니 왜이렇게 코드를 작성했을까 라는 생각이 잘 안드는거 같아요. 
요구사항에서 왜 이렇게 짜라고 했을까, 더 좋은 방법이 있을까 등등 생각해보면서 코드를 작성한 이유에 대해 더 생각해보겠습니다.

kimnayeon0108 committed:
Fix distinguish() method
rename it to getPieceList, more clear what the method does
change location of getPieceList() method, make it easier when code is read

kimnayeon0108 committed:
Remove size() method
for the reason of being able to get size of piece lists in the class and there is no case to get size of piece lists out of the class

kimnayeon0108 committed:
Remove findPawn() method
called nowhere

kimnayeon0108 committed:
Rename canstant PAWNSNUM to BOARD_RAND
board has eight rank

kimnayeon0108 committed:
Change shoBoard() method
replace using String appending function with + as using StringBuilder with append method

kimnayeon0108 committed:
Fix Chess class
split method to divide function
add getInput() method, startChess() method

kimnayeon0108 committed:
Add DisplayName annotation to clerify what check() method does

kimnayeon0108 commented:
@honux77 코드 수정했으니 리뷰 부탁드립니다! 
size()와 findPawn()는 이전 미션에서 구현한 것이라서 사용하지 않아도 삭제하지 않고 뒀는데 고민하다가 나중에 필요할때 적절히 다시 구현하는게 나을거같아서 삭제했습니다.

Chess 클래스에서 사용자 입력받아서 시작하는 부분은 try-with-resources로 구현해보려고 했는데 catch 구문에서 다시 입력받도록 하는 것을 구현하지 못했습니다.  try의 괄호'( )' 안에 Scanner 객체를 생성하게되면 catch구문에서 try 괄호 안에 스캐너 변수에 접근하지 못해서 다시 입력받는걸 구현하는데 실패해서 일단 close()를 추가해주었습니다.

honux77 commented:
수고하셨어요. close() 해줘도 됩니다.
